src/libraries/Common/src/System/TimeProvider.cs

```diff
@@ -0,0 +1,207 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System
+{
+    /// <summary>Provides an abstraction for time.</summary>
+    public abstract class TimeProvider
+    {
+        private readonly double _timeToTicksRatio;
+
+        /// <summary>
+        /// Gets a <see cref="TimeProvider"/> that provides a clock based on <see cref="DateTimeOffset.UtcNow"/>,
+        /// a time zone based on <see cref="TimeZoneInfo.Local"/>, a high-performance time stamp based on <see cref="Stopwatch"/>,
+        /// and a timer based on <see cref="Timer"/>.
+        /// </summary>
+        public static TimeProvider System { get; } = new SystemTimeProvider(null);
+
+        /// <summary>
+        /// Initializes the instance with the timestamp frequency.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">The value of <paramref name="timestampFrequency"/> is negative or zero.</exception>
+        /// <param name="timestampFrequency">Frequency of the values returned from <see cref="GetTimestamp"/> method.</param>
+        protected TimeProvider(long timestampFrequency)
+        {
+            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(timestampFrequency);
+            TimestampFrequency = timestampFrequency;
+            _timeToTicksRatio = (double)TimeSpan.TicksPerSecond / TimestampFrequency;
+        }
+
+        /// <summary>
+        /// Gets a <see cref="DateTimeOffset"/> value whose date and time are set to the current
+        /// Coordinated Universal Time (UTC) date and time and whose offset is Zero,
+        /// all according to this <see cref="TimeProvider"/>'s notion of time.
+        /// </summary>
+        public abstract DateTimeOffset UtcNow { get; }
+
+        /// <summary>
+        /// Gets a <see cref="DateTimeOffset"/> value that is set to the current date and time according to this <see cref="TimeProvider"/>'s
+        /// notion of time based on <see cref="UtcNow"/>, with the offset set to the <see cref="LocalTimeZone"/>'s offset from Coordinated Universal Time (UTC).
+        /// </summary>
+        public DateTimeOffset LocalNow
+        {
+            get
+            {
+                DateTime utcDateTime = UtcNow.UtcDateTime;
+                TimeSpan offset = LocalTimeZone.GetUtcOffset(utcDateTime);
+
+                long localTicks = utcDateTime.Ticks + offset.Ticks;
+                if ((ulong)localTicks > DateTime.MaxTicks)
+                {
+                    localTicks = localTicks < DateTime.MinTicks ? DateTime.MinTicks : DateTime.MaxTicks;
+                }
+
+                return new DateTimeOffset(localTicks, offset);
```

Comment:
Why not using `TimeZoneInfo.ConvertTime(DateTimeOffset, TimeZoneInfo)`? Is there any behavior difference?

`DateTimeOffset.ToOffset` can throw exception when the local time is out of range.
Converting the zero point of time to negative offset has always been an issue. We should at least clarify it in documentation.

Reply:
> Why not using TimeZoneInfo.ConvertTime(DateTimeOffset, TimeZoneInfo)? Is there any behavior difference?

Just for perf. The behavior should be the same. 

> Converting the zero point of time to negative offset has always been an issue. 

Note this only can happen if someone overrides UtcNow to return a very-small value (like DateTimeOffset.MinValue) with negative offset. I believe throwing in such a case will be desirable because this can only happen in test scenarios. I cannot imagine anyone can do that in the production scenario.